File: netty/src/main/java/com/townwang/netty/NettyClientHandler.java
/*
 * Copyright © 文科中的技术宅
 * blog:https://www.townwang.com
 */
package com.townwang.netty;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;

/**
 * @author Town
 * @created at 2018/7/11 18:28
 * @Last Modified by: Town
 * @Last Email: android@townwang.com
 * @Last Modified time: 2018/7/11 18:28
 * @Remarks
 */
public class NettyClientHandler extends SimpleChannelInboundHandler<ByteBuf> {

    /**
     * 回调接口
     */
    private NettyListener listener;

    public NettyClientHandler(NettyListener listener) {
        this.listener = listener;
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) {
        try {
            NeetyTown.getInstance().setConnectStatus(true);
            listener.onServiceStatusConnectChanged(NettyListener.STATUS_CONNECT_SUCCESS);
        } catch (Exception e) {
            e.getMessage();
        }
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) {
        try {
            NeetyTown.getInstance().setConnectStatus(false);
            listener.onServiceStatusConnectChanged(NettyListener.STATUS_CONNECT_CLOSED);
            NeetyTown.getInstance().reconnect();
        } catch (Exception e) {
            e.getMessage();
        }
    }

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, ByteBuf byteBuf) {
//        Logger.d("Netty TCP ---> 执行这里了吗"+byteBuf);
        try {
            byte[] req = new byte[byteBuf.readableBytes()];
            byteBuf.readBytes(req);
            String body = new String(req, "UTF-8");
            listener.onMessageResponse(body);
        } catch (Exception e) {
            e.getMessage();
        }
    }
}

File: app/src/main/java/com/townwang/tcp/tcp/NettyClientHandler.kt
package com.townwang.tcp.tcp
import android.util.Log
import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler

class NettyClientHandler(

    /**
     * 回调接口
     */
    private val listener: NettyListener
) : SimpleChannelInboundHandler<ByteBuf>() {
    private val  TAG = "TCP-netty"

    override fun channelActive(ctx: ChannelHandlerContext) {
        try {
            MyNettyClitent.instance.connectStatus = true
            listener.onServiceStatusConnectChanged(NettyListener.STATUS_CONNECT_SUCCESS.toInt())
        } catch (e: Exception) {
            e.message
        }

    }

    override fun channelInactive(ctx: ChannelHandlerContext) {
        try {
            MyNettyClitent.instance.connectStatus = false
            listener.onServiceStatusConnectChanged(NettyListener.STATUS_CONNECT_CLOSED.toInt())
            MyNettyClitent.instance.reconnect()
        } catch (e: Exception) {
            e.message
        }

    }

    override fun channelRead0(ctx: ChannelHandlerContext, byteBuf: ByteBuf) {
        try {
            val req = ByteArray(byteBuf.readableBytes())
            byteBuf.readBytes(req)
            val body = String(req)
            if (body.contains("HEARTBEAT")) {
                Log.d(TAG,"Netty TCP 收到心跳包 ---> 【$body】 ")
            } else {
                Log.d(TAG,"Netty TCP 收到消息 ---> 【$body】 ")
                listener.onMessageResponse(body)
            }
        } catch (e: Exception) {
            e.message
        }

    }

}
